torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -2661,4 +2677,73 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
   return work;
 }
 
+void ProcessGroupGloo::monitoredBarrier(const BarrierOptions& opts) {
+  auto monitoredBarrierTimeout = opts.timeout;
+  auto rank = this->getRank();
+  auto t1 = nextTag();
+  auto t2 = nextTag();
+  std::vector<at::Tensor> t = {at::tensor({rank})};
+  // only enforce timeout on rank 0. This is so that other ranks aren't timed
+  // out first, bringing down the job without reporting which rank timed out.
+  if (rank != 0) {
+    auto sendWork = send(t, 0, t1);
+    sendWork->wait();
+    auto recvWork = recv(t, 0, t2);
+    recvWork->wait();
+    return;
+  }
+  auto startTime = std::chrono::steady_clock::now();
+  auto worldSize = this->getSize();
+  for (int rank = 0; rank < worldSize; ++rank) {
+    // Rank 0 recv
+    if (rank != 0) {
+      auto recvWork = recv(t, rank, t1);
+      auto remainingTime = monitoredBarrierTimeout == kUnsetTimeout
```

Comment:
Why do we use `getRemainingtime` here? We can just pass `monitoredBarrierTimeout` to `wait`?

Reply:
Do you mean we can omit the call to `getRemainingTime` for the very first call to the collective, or in general why do we use `getRemainingTime` at all? 

For the latter, I implemented it this way so that if timeout = 100s for example, then we wait a total of 100s and not 100s for each call to a collective work's wait(). For the former point, I guess we can omit the call for the very first collective since all the time is remaining.